### PR TITLE
[Velocity smoother] Set zeros command if timeout

### PR DESCRIPTION
**Pull request by BriceRenaudeau**
## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #3481 |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | gazebo simulation |

---

## Description of contribution in a few bullet points
 In the timeout check :
* I set the `command_` to zero
* I check if the `last_cmd_` was zero to set the `stopped_` flag to true to stop publishing.

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

**Comment by BriceRenaudeau** [comment]
CI test failed on an other node.
Might not be related.
Please rerun.

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: nav2_velocity_smoother/src/velocity_smoother.cpp [hunk 1]
```diff
@@ -216,14 +216,13 @@ void VelocitySmoother::smootherTimer()
...
 
-  // Check for velocity timeout. If nothing received, publish zeros to stop robot
+  // Check for velocity timeout. If nothing received, publish zeros to apply deceleration
   if (now() - last_command_time_ > velocity_timeout_) {
-    last_cmd_ = geometry_msgs::msg::Twist();
-    if (!stopped_) {
-      smoothed_cmd_pub_->publish(std::move(cmd_vel));
+    if (last_cmd_ == geometry_msgs::msg::Twist() || stopped_) {
+      stopped_ = true;
+      return;
```
[the remark this answers is not in the document]
Comment (BriceRenaudeau) [reply nav2_velocity_smoother/src/velocity_smoother.cpp:223]: When you enter the timeout, the robot was moving (else it's already stopped) so the `last_cmd_` is different from zeros.
You don't enter the second if. I set the command to zero as if it received the command hence it applies the deceleration correctly. 
You reach the `return` only if the `last_cmd_` is zero which means the deceleration reached zero.

I tested it and it works well.

[the remark this answers is not in the document]
Comment (SteveMacenski) [reply nav2_velocity_smoother/src/velocity_smoother.cpp:223]: I'm looking at a few timesteps of this running:
- We hit a timeout where we receive no new data
- We set the command to a null value and continue on as usual with a `0`ed out velocity. That's set to last command at the end.
- The next round, we're still not receiving new data so we find that last command is zero so we set stopped to `true` and return
- From that point on, we return continuously and never actually enter the body of the function to publish any new commands to apply this node's functionality. So as far as I read this, any such deceleration is just the outcome of your mobile base's handling, not due to this node winding down the velocity profile by the set acceleration limits

This tells me that this is only called for 2 timer cycles after the commands stop coming in, as opposed to 1 before.

[the remark this answers is not in the document]
Comment (SteveMacenski) [reply nav2_velocity_smoother/src/velocity_smoother.cpp:223]: You are correct, there are many relatedly named things I wrote in here I got wrapped up in. I agree now that this does that you describe.

**Review by SteveMacenski — APPROVED** [review]

**Comment by BriceRenaudeau** [comment]
I made a PR to add an entry in the docs:
https://github.com/ros-planning/navigation.ros.org/pull/409

**Comment by SteveMacenski** [comment]
Thanks!

**Comment by ufrhaidar** [comment]
why is this `last_cmd_ = *cmd_vel;` moved after the deadband?

If you have a deadband in openloop the machine will get stuck at 0 commanded velocity.

@SteveMacenski @BriceRenaudeau

**Comment by SteveMacenski** [comment]
The order shouldn't matter since `last_cmd_` isn't used in between. From before, `if (open_loop_) {` still set the same value as such, so there should be absolutely no change in behavior

**Comment by ufrhaidar** [comment]
> The order shouldn't matter since `last_cmd_` isn't used in between. From before, `if (open_loop_) {` still set the same value as such, so there should be absolutely no change in behavior

It does. If:
- current velocity = `0.0`
- commanded velocity = `0.5`
- max acceleration = `1.0`
- deadband = `0.2`

`applyConstraints()` will make `cmd_vel` be equal to `0.05`, the deadband will make it `0.0`, and then `last_cmd_` will be `0.0`.

You'll end up the same place you started and will get stuck in a loop.

**Comment by SteveMacenski** [comment]
@BriceRenaudeau can you comment? I think @ufrhaidar has a point

**Comment by BriceRenaudeau** [comment]
I will check the configuration tomorrow.

My thoughts is that in open loop the previous command is used, so increase at every step and is not stuck. The deadband is only applied to real published command, not the saved one.

**Comment by BriceRenaudeau** [comment]
There may be an issue.

The line:
https://github.com/ros-planning/navigation2/blob/778c0faee41788c9e3323a420ff2c10122ba4aae/nav2_velocity_smoother/src/velocity_smoother.cpp#L314

Must be above the deadband test, not below.
As I said, I will test that tomorrow.

**Comment by SteveMacenski** [comment]
If you’re happy open a followup PR to move it and I’ll merge!

**Comment by BriceRenaudeau** [comment]
I was able to reproduce the issue. The velocity smoother was stuck with a hight deadband (0.2).

The fix is working correctly.

**Comment by SteveMacenski** [comment]
Thanks for the note!

**Pull request merged**